File: services/followers/src/services/notifier.js
const config = require("../config");

// Best-effort: a notification failing (stakeholders down/slow) should never
// break the follow/unfollow request that triggered it.
async function notify(username, type, message, relatedUsername) {
  try {
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), 3000);
    try {
      await fetch(`${config.stakeholdersUrl}/stakeholders/notifications/internal`, {
        method: "POST",
        // Connection: close - stakeholders can run as multiple replicas
        // behind Docker's embedded DNS (docker-compose.yml no longer pins
        // its container_name). Node's fetch keeps HTTP connections alive
        // and reuses them by default, which would pin this call to
        // whichever replica answered first; closing after each call forces
        // a fresh connection (and DNS lookup) so these low-volume
        // notifications actually spread across replicas.
        headers: { "Content-Type": "application/json", "Connection": "close" },
        body: JSON.stringify({ username, type, message, relatedUsername }),
        signal: controller.signal,
      });
    } finally {
      clearTimeout(timeout);
    }
  } catch (err) {
    console.error(JSON.stringify({ level: "warn", msg: "notification failed", error: String(err) }));
  }
}

async function notifyNewFollower(followerUsername, targetUsername) {
  await notify(
    targetUsername,
    "follow",
    `${followerUsername} je počeo/la da te prati.`,
    followerUsername
  );
}

module.exports = { notifyNewFollower };

File: services/tours/notify/notifier.go
// Package notify makes a best-effort call to stakeholders' internal
// notification endpoint. A failure here (stakeholders down/slow) must never
// break whatever triggered the notification.
package notify

import (
	"bytes"
	"encoding/json"
	"log/slog"
	"net/http"
	"os"
	"time"
)

var client = &http.Client{Timeout: 3 * time.Second}

func stakeholdersURL() string {
	url := os.Getenv("STAKEHOLDERS_URL")
	if url == "" {
		url = "http://localhost:8081"
	}
	return url
}

type payload struct {
	Username        string `json:"username"`
	Type            string `json:"type"`
	Message         string `json:"message"`
	RelatedUsername string `json:"relatedUsername,omitempty"`
}

func send(p payload) {
	body, err := json.Marshal(p)
	if err != nil {
		return
	}
	resp, err := client.Post(stakeholdersURL()+"/stakeholders/notifications/internal", "application/json", bytes.NewReader(body))
	if err != nil {
		slog.Warn("notification delivery failed", "username", p.Username, "error", err)
		return
	}
	defer resp.Body.Close()
}

// NewReview notifies tourAuthorUsername that touristUsername reviewed
// their tour. Fire-and-forget - call it in a goroutine.
func NewReview(tourAuthorUsername, touristUsername string) {
	if tourAuthorUsername == "" || tourAuthorUsername == touristUsername {
		return
	}
	go send(payload{
		Username:        tourAuthorUsername,
		Type:            "review",
		Message:         touristUsername + " je ostavio/la recenziju na tvoju turu.",
		RelatedUsername: touristUsername,
	})
}
